typed_sql/lib/src/typed_sql.query.dart

```diff
@@ -0,0 +1,395 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+part of 'typed_sql.dart';
+
+/// A [Query] on the database from which results can be fetched.
+final class Query<T extends Record> {
+  final Database _context;
+
+  final T _expressions;
+  final QueryClause Function(List<Expr> expressions) _from;
+
+  Query._(this._context, this._expressions, this._from);
+
+  // TODO: Consider a toString method!
+}
+
+final class InnerJoin<T extends Record, S extends Record> {
+  final Query<T> _from;
+  final Query<S> _join;
+
+  InnerJoin._(this._from, this._join);
+}
+
+final class LeftJoin<T extends Record, S extends Record> {
+  final Query<T> _from;
+  final Query<S> _join;
+
+  LeftJoin._(this._from, this._join);
+}
+
+final class RightJoin<T extends Record, S extends Record> {
+  final Query<T> _from;
+  final Query<S> _join;
+
+  RightJoin._(this._from, this._join);
+}
+
+/*
+final class View<T extends Record> extends Query<T> {
+  @override
+  T get _expressions => throw UnimplementedError();
+
+  @override
+  QueryClause Function(List<Expr<Object?>> expressions) get _from =>
+      throw UnimplementedError();
+
+  View._(super._context) : super._();
+}*/
+
+final class Table<T extends Model> extends Query<(Expr<T>,)> {
+  final TableClause _tableClause;
+
+  Table._(
+    Database context,
+    this._tableClause,
+    TableDefinition<T> definition,
+  ) : super._(
+          context,
+          (ModelExpression._(0, definition, Object()),),
+          (_) => _tableClause,
+        );
+}
+
+final class QuerySingle<T extends Record> {
+  final Query<T> _query;
+
+  QuerySingle._(this._query);
+}
+
+/// A [Query] which can only be used as a subquery.
+final class SubQuery<T extends Record> {
+  final T _expressions;
+  final QueryClause Function(List<Expr> expressions) _from;
+
+  SubQuery._(this._expressions, this._from);
+}
+
+/* --------------------- Query clauses ---------------------- */
+
+sealed class QueryClause {}
+
+final class TableClause extends QueryClause {
+  final TableDefinition _definition;
+
+  /// Name of table
+  String get name => _definition.tableName;
+  List<String> get columns => _definition.columns;
+  List<String> get primaryKey => _definition.primaryKey;
+
+  TableClause._(this._definition);
+}
+
+final class SelectClause extends QueryClause {
+  final List<Expr> _expressions;
+
+  Iterable<Expr> get expressions => _expressions.expand((e) => e._explode());
+
+  SelectClause._(this._expressions);
+}
+
+/*
+final class ViewClause extends QueryClause {
+  /// Name of view
+  final String name;
+  final List<String> columns;
+  ViewClause(this.name, this.columns);
+}*/
+
+sealed class FromClause extends QueryClause {
+  final QueryClause from;
+  FromClause._(this.from);
+}
+
+/// Interface implemented by object with-in which expressions may exist.
+///
+/// Expressions can be bound to context, that is the context from which they
+/// are referencing fields.
```

Comment:
This is hard to read...

Reply:
Also not part of the public API, necessary for implementing a dialect, which is not something we declare stable anytime soon.